### PR TITLE
[SPARK-26686][SS] Remove unnecessary KafkaSourceProvider parameter lowercase conversion

### PR DESCRIPTION
**Pull request by gaborgsomogyi**
## What changes were proposed in this pull request?

KafkaSourceProvider receives lowercase parameters but it converts them again which is unnecessary.

In this PR I've removed these conversions.

## How was this patch tested?

Existing unit tests.

### CODE REVIEW

**Comment by gaborgsomogyi** [comment]
cc @HeartSaVioR @zsxwing

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala [hunk 1]
```diff
@@ -121,19 +120,13 @@ private[kafka010] class KafkaSourceProvider extends DataSourceRegister
...
     val uniqueGroupId = streamingUniqueGroupId(parameters, metadataPath)
 
-    val caseInsensitiveParams = parameters.map { case (k, v) => (k.toLowerCase(Locale.ROOT), v) }
-    val specifiedKafkaParams =
-      parameters
-        .keySet
-        .filter(_.toLowerCase(Locale.ROOT).startsWith("kafka."))
-        .map { k => k.drop(6).toString -> parameters(k) }
-        .toMap
+    val specifiedKafkaParams = convertToSpecifiedParams(parameters)
```
Comment (gaborgsomogyi) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala]: This change is unrelated but didn't want to leave it there.

**Comment by gaborgsomogyi** [comment]
Either `DataSource` pass `caseInsensitiveOptions` as a parameter or `DataSourceOptions.asMap` used. Both provides lowercase keys.

**Comment by HeartSaVioR** [comment]
IMHO I'm a bit afraid (at least not 100% sure) that CaseInsensitiveMap will provide lowercase keys at all cases. (I'm not an expert of Scala honestly.) I just read MapLike and realize many of other methods rely on `get` and `iterator` so safe to assume they'll provide lowercase key as well, but not sure I'm not missing other methods.

**Comment by gaborgsomogyi** [comment]
> Moreover, we don't guarantee providing key-lowercased map in interface level (though Spark provides CaseInsensitiveMap for now).

Here are the interfaces which are involved:
- DataSourceOptions:
```
 * An immutable string-to-string map in which keys are case-insensitive. This is used to represent
 * data source options.
```
- RelationProvider:
```
   * @note The parameters' keywords are case insensitive and this insensitivity is enforced
   * by the Map that is passed to the function.
```
- BaseRelation: Here the case-insensitivity only seen from usage
- StreamSourceProvider: Same here
- CreatableRelationProvider: Same here

All in all tend to agree with you. Even if the actual situation is case-insensitive params are arriving it's not enforced on the language/interface level which could cause issues later.

I suggest to remove conversion where `DataSourceOptions` passed. The reasoning behind that this class is designed exactly this way (many features are depending on this). The rest where a simple map provided can stay to be defensive.

**Comment by HeartSaVioR** [comment]
> An immutable string-to-string map in which keys are case-insensitive.

Suppose we are only aware of the javadoc, then actually this doesn't mean map will provide key as lowercase. Case-insensitive can be achieved with various approaches and the map can still provide original key and value, and even uppercase. I may be unnecessarily defensive about that, but just seeing this as other connector implementors' perspective.

Let's hear other opinions as well.

**Comment by jose-torres** [comment]
I'm not sure I understand why this change is valuable, but I agree that it looks valid, and I'd argue that DataSourceOptions can/does/should guarantee it canonicalizes the case-insensitive map to something that implementations can use as a raw Map[String, String] if they want.

Are there existing tests which verify that parameters are passed through to Kafka case-insensitively? I'm not necessarily comfortable just trusting generic unit tests, since failing to pass parameters from our test setup might not cause observable problems.

**Comment by srowen** [comment]
So is the conclusion here that the case-insensitive map doesn't necessarily provide normalized keys, and that's a problem that also needs to be addressed? or not a problem? I am not as familiar with the code, and wasn't clear from the discussion here what the conclusion was.

**Comment by HeartSaVioR** [comment]
To be short, what I concerned is guaranteeing case-insensitive is not same as guaranteeing lower keys (I'm not also 100% feeling safe on 'normalized' since it can be also done with upper keys). I've seen javadoc in some interfaces guaranteed former, but not latter. IMHO to remove the conversion (which the patch is trying to achieve), the interfaces need to guarantee latter, or Kafka should be able to handle the parameters by case-insensitive manner.

**Comment by gaborgsomogyi** [comment]
What I would like to do here is to do order because there are some code line which do validation without making things lowecase for example [this](https://github.com/apache/spark/blob/e97ab1d9807134bb557ae73920af61e8534b2b08/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala#L66), [this](https://github.com/apache/spark/blob/e97ab1d9807134bb557ae73920af61e8534b2b08/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala#L81) and [this](https://github.com/apache/spark/blob/e97ab1d9807134bb557ae73920af61e8534b2b08/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala#L118). I'm fine with moving the mentioned codeparts after the lowercase conversion but what I'm sure we need to do something because part of the parameter handling is depending on the (maybe wrong) fact that it's already lowered. Opinions?

**Comment by gaborgsomogyi** [comment]
FYI, I think we have bigger issues in this area. I've just filed [SPARK-26783](https://issues.apache.org/jira/browse/SPARK-26783).

**Comment by gaborgsomogyi** [comment]
As the API is not explicitly stating that it provides lowercase parameters better to keep sometimes double conversion for safety reasons.

The bad part is there are code lines which are depending on the assumption that these APIs providing lowercase parameters.

**Pull request closed without merging**